File: Assets/Bundles/Shared/Resources/TS/UserInput/Mouse.lua
-- Compiled with unity-ts v2.1.0-75
local Bin = require("Shared/TS/Util/Bin").Bin
local Signal = require("Shared/TS/Util/Signal").Signal
local MouseDriver = require("Shared/TS/UserInput/Drivers/MouseDriver").MouseDriver
local mouseUnlockerKeys = {}
local mouseUnlockerIdCounter = 1
local Mouse
do
	Mouse = setmetatable({}, {
		__tostring = function()
			return "Mouse"
		end,
	})
	Mouse.__index = Mouse
	function Mouse.new(...)
		local self = setmetatable({}, Mouse)
		return self:constructor(...) or self
	end
	function Mouse:constructor()
		self.bin = Bin.new()
		self.mouseDriver = MouseDriver:instance()
		self.LeftDown = Signal.new()
		self.LeftUp = Signal.new()
		self.RightDown = Signal.new()
		self.RightUp = Signal.new()
		self.MiddleDown = Signal.new()
		self.MiddleUp = Signal.new()
		self.Scrolled = Signal.new()
		self.Moved = Signal.new()
		self.Delta = Signal.new()
		self.isLeftDown = false
		self.isRightDown = false
		self.isMiddleDown = false
		self.location = Vector3.new(0, 0, 0)
		-- Track signals in bin:
		self.bin:Add(self.LeftDown)
		self.bin:Add(self.LeftUp)
		self.bin:Add(self.RightDown)
		self.bin:Add(self.RightUp)
		self.bin:Add(self.LeftDown)
		self.bin:Add(self.LeftUp)
		self.bin:Add(self.Scrolled)
		self.bin:Add(self.Moved)
		self.bin:Add(self.Delta)
		-- Initial states:
		self.isLeftDown = self.mouseDriver:IsLeftDown()
		self.isRightDown = self.mouseDriver:IsRightDown()
		self.isMiddleDown = self.mouseDriver:IsMiddleDown()
		self.location = self.mouseDriver:GetLocation()
		-- ▼ ReadonlySet.size ▼
		local _size = 0
		for _ in mouseUnlockerKeys do
			_size += 1
		end
		-- ▲ ReadonlySet.size ▲
		if _size == 0 then
			self.mouseDriver:SetLocked(true)
		end
		-- Connect to mouse driver:
		self.bin:Connect(self.mouseDriver.LeftButtonChanged, function(isDown)
			self.isLeftDown = isDown
			if isDown then
				self.LeftDown:Fire()
			else
				self.LeftUp:Fire()
			end
		end)
		self.bin:Connect(self.mouseDriver.RightButtonChanged, function(isDown)
			self.isRightDown = isDown
			if isDown then
				self.RightDown:Fire()
			else
				self.RightUp:Fire()
			end
		end)
		self.bin:Connect(self.mouseDriver.MiddleButtonChanged, function(isDown)
			self.isMiddleDown = isDown
			if isDown then
				self.MiddleDown:Fire()
			else
				self.MiddleUp:Fire()
			end
		end)
		self.bin:Connect(self.mouseDriver.Moved, function(location)
			self.location = location
			self.Moved:Fire(location)
		end)
		self.bin:Connect(self.mouseDriver.Delta, function(delta)
			self.Delta:Fire(delta)
		end)
		self.bin:Connect(self.mouseDriver.Scrolled, function(delta)
			self.Scrolled:Fire(delta)
		end)
	end
	function Mouse:IsLeftButtonDown()
		return self.isLeftDown
	end
	function Mouse:IsRightButtonDown()
		return self.isRightDown
	end
	function Mouse:IsMiddleButtonDown()
		return self.isMiddleDown
	end
	function Mouse:GetLocation()
		return self.location
	end
	function Mouse:SetLocation(position)
		self.mouseDriver:SetLocation(position)
	end
	function Mouse:GetDelta()
		return self.mouseDriver:GetDelta()
	end
	function Mouse:AddUnlocker()
		local id = mouseUnlockerIdCounter
		mouseUnlockerIdCounter += 1
		self.mouseDriver:SetLocked(false)
		return id
	end
	function Mouse:RemoveUnlocker(id)
		local _id = id
		mouseUnlockerKeys[_id] = nil
		-- ▼ ReadonlySet.size ▼
		local _size = 0
		for _ in mouseUnlockerKeys do
			_size += 1
		end
		-- ▲ ReadonlySet.size ▲
		if _size == 0 then
			self.mouseDriver:SetLocked(true)
		end
	end
	function Mouse:ClearAllLockers()
		table.clear(mouseUnlockerKeys)
		self.mouseDriver:SetLocked(true)
	end
	function Mouse:IsLocked()
		return self.mouseDriver:IsLocked()
	end
	function Mouse:Destroy()
		self.bin:Destroy()
	end
end
return {
	Mouse = Mouse,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/UserInput/Mouse.ts
import { Bin } from "Shared/Util/Bin";
import { Signal } from "Shared/Util/Signal";
import { MouseDriver } from "./Drivers/MouseDriver";

const mouseUnlockerKeys = new Set<number>();
let mouseUnlockerIdCounter = 1;

export class Mouse {
	private readonly bin = new Bin();
	private readonly mouseDriver = MouseDriver.instance();

	public readonly LeftDown = new Signal<void>();
	public readonly LeftUp = new Signal<void>();
	public readonly RightDown = new Signal<void>();
	public readonly RightUp = new Signal<void>();
	public readonly MiddleDown = new Signal<void>();
	public readonly MiddleUp = new Signal<void>();
	public readonly Scrolled = new Signal<[delta: number]>();
	public readonly Moved = new Signal<[location: Vector3]>();
	// public readonly Delta = new Signal<[delta: Vector3]>();

	private isLeftDown = false;
	private isRightDown = false;
	private isMiddleDown = false;
	private location = new Vector3(0, 0, 0);

	constructor() {
		// Track signals in bin:
		this.bin.Add(this.LeftDown);
		this.bin.Add(this.LeftUp);
		this.bin.Add(this.RightDown);
		this.bin.Add(this.RightUp);
		this.bin.Add(this.LeftDown);
		this.bin.Add(this.LeftUp);
		this.bin.Add(this.Scrolled);
		this.bin.Add(this.Moved);
		// this.bin.Add(this.Delta);

		// Initial states:
		this.isLeftDown = this.mouseDriver.IsLeftDown();
		this.isRightDown = this.mouseDriver.IsRightDown();
		this.isMiddleDown = this.mouseDriver.IsMiddleDown();
		this.location = this.mouseDriver.GetLocation();

		if (mouseUnlockerKeys.size() === 0) {
			this.mouseDriver.SetLocked(true);
		}

		// Connect to mouse driver:

		this.bin.Connect(this.mouseDriver.LeftButtonChanged, (isDown) => {
			this.isLeftDown = isDown;
			if (isDown) {
				this.LeftDown.Fire();
			} else {
				this.LeftUp.Fire();
			}
		});

		this.bin.Connect(this.mouseDriver.RightButtonChanged, (isDown) => {
			this.isRightDown = isDown;
			if (isDown) {
				this.RightDown.Fire();
			} else {
				this.RightUp.Fire();
			}
		});

		this.bin.Connect(this.mouseDriver.MiddleButtonChanged, (isDown) => {
			this.isMiddleDown = isDown;
			if (isDown) {
				this.MiddleDown.Fire();
			} else {
				this.MiddleUp.Fire();
			}
		});

		this.bin.Connect(this.mouseDriver.Moved, (location) => {
			this.location = location;
			this.Moved.Fire(location);
		});

		// this.bin.Connect(this.mouseDriver.Delta, (delta) => {
		// 	this.Delta.Fire(delta);
		// });

		this.bin.Connect(this.mouseDriver.Scrolled, (delta) => {
			this.Scrolled.Fire(delta);
		});
	}

	/** Returns `true` if the left mouse button is down. */
	public IsLeftButtonDown() {
		return this.isLeftDown;
	}

	/** Returns `true` if the right mouse button is down. */
	public IsRightButtonDown() {
		return this.isRightDown;
	}

	/** Returns `true` if the middle mouse button is down. */
	public IsMiddleButtonDown() {
		return this.isMiddleDown;
	}

	/** Gets the position of the mouse on-screen. */
	public GetLocation() {
		return this.location;
	}

	/** Sets the position of the mouse on-screen. */
	public SetLocation(position: Vector3) {
		this.mouseDriver.SetLocation(position);
	}

	/** Gets the mouse's change in position on-screen over the last frame. */
	public GetDelta() {
		return this.mouseDriver.GetDelta();
	}

	/**
	 * Locks the mouse.
	 * Returns an ID that can be used to unlock the mouse.
	 */
	public AddUnlocker(): number {
		const id = mouseUnlockerIdCounter;
		mouseUnlockerIdCounter++;

		this.mouseDriver.SetLocked(false);

		return id;
	}

	public RemoveUnlocker(id: number): void {
		mouseUnlockerKeys.delete(id);
		if (mouseUnlockerKeys.size() === 0) {
			this.mouseDriver.SetLocked(true);
		}
	}

	public ClearAllLockers(): void {
		mouseUnlockerKeys.clear();
		this.mouseDriver.SetLocked(true);
	}

	/** Check if the mouse is locked. */
	public IsLocked() {
		return this.mouseDriver.IsLocked();
	}

	/** Cleans up the mouse. */
	public Destroy() {
		this.bin.Destroy();
	}
}
